File: content/snippets/python/problems_file_aggregate.py
#!/usr/bin/env python3
from dataclasses import dataclass
from typing import List, Dict
from concurrent.futures import ThreadPoolExecutor
import statistics
import sys

@dataclass
class Record:
    name: str
    score: int

class FileProcessor:
    def __init__(self, path: str):
        self.path = path

    def read_all(self) -> List[Record]:
        records: List[Record] = []
        with open(self.path, "r", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                try:
                    name, s = [x.strip() for x in line.split(",", 1)]
                    score = int(s)
                    records.append(Record(name, score))
                except Exception:
                    continue
        return records


def average(scores: List[int]) -> float:
    return statistics.mean(scores) if scores else 0.0


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else "data.txt"
    fp = FileProcessor(path)
    try:
        records = fp.read_all()
        buckets: Dict[str, List[int]] = {}
        for r in records:
            buckets.setdefault(r.name, []).append(r.score)

        results = []
        with ThreadPoolExecutor() as ex:
            futures = {ex.submit(average, scores): name for name, scores in buckets.items()}
            for fut in futures:
                name = futures[fut]
                results.append((name, fut.result()))

        results.sort(key=lambda x: x[1], reverse=True)
        for name, avg in results:
            print(f"{name}: {avg:.2f}")
    except Exception as e:
        print("Error:", e)


if __name__ == "__main__":
    main()

File: content/snippets/typescript/problems_file_aggregate.ts
#!/usr/bin/env ts-node
import * as fs from "fs/promises";

type Record = { name: string; score: number };

class FileProcessor {
    path: string;
    constructor(path: string) {
        this.path = path;
    }

    async readAll(): Promise<Record[]> {
        const text = await fs.readFile(this.path, { encoding: "utf8" });
        const out: Record[] = [];
        for (const raw of text.split(/\r?\n/)) {
            const line = raw.trim();
            if (!line || line.startsWith("#")) continue;
            const parts = line.split(",", 2);
            if (parts.length < 2) continue;
            const name = parts[0].trim();
            const score = Number(parts[1].trim());
            if (Number.isNaN(score)) continue;
            out.push({ name, score });
        }
        return out;
    }
}

function average(nums: number[]): number {
    if (nums.length === 0) return 0;
    return nums.reduce((a, b) => a + b, 0) / nums.length;
}

async function main() {
    const path = process.argv[2] || "data.txt";
    try {
        const fp = new FileProcessor(path);
        const records = await fp.readAll();
        const buckets = new Map<string, number[]>();
        for (const r of records) {
            if (!buckets.has(r.name)) buckets.set(r.name, []);
            buckets.get(r.name)!.push(r.score);
        }

        const promises: Promise<[string, number]>[] = [];
        for (const [name, scores] of buckets.entries()) {
            promises.push(Promise.resolve([name, average(scores)]));
        }
        const results = (await Promise.all(promises)).sort((a, b) => b[1] - a[1]);
        for (const [name, avg] of results) {
            console.log(`${name}: ${avg.toFixed(2)}`);
        }
    } catch (err) {
        console.error("Error:", err);
        process.exit(1);
    }
}

main();
